src/vm/clsload.cpp

```diff
@@ -3186,11 +3187,27 @@ TypeHandle ClassLoader::DoIncrementalLoad(TypeKey *pTypeKey, TypeHandle typeHnd,
         case CLASS_LOAD_BEGIN :
             {
                 IBCLoggerAwareAllocMemTracker amTracker;
+
+                // A failure in PublishType effectively "unloads" types that failed to load.
+                // In particular the same addresses may be used by different types in the future.
+                // There is a remote possibility that type loading itself may indirectly cause casting 
+                // analysis and may end up caching results for rejected types.
+                // To be sure that, once new types are created, there are no pre-filled incorrect entries, 
+                // we will flush the cast cache on type load failures.
+                // RE: perf concerns. Type load rejections, especially survivable ones is very rare scenario.
+                // RE: concurrency. Type publishing moves a volatile pointer. Anyone seeing a new type will see flushed cache.
```

Comment:
I do not think so. They are common, in particular in big complex apps. For example, it is common to fail type loading due to missing dependencies.

It would be better to avoid storing any partially loaded types into the cache. I believe that the casting should not ever run on partially loaded types anyway. You can both assert that whatever we are storing into the cache is fully loaded - it would be useful to find out whether there are cases where the casting runs on partially loaded types; and also skip it.